docs/CNDP_README.md

```diff
@@ -0,0 +1,371 @@
+        --cpuset-cpus=22-25 --ulimit memlock=-1 --cap-add IPC_LOCK \
+        -v /dev/hugepages:/dev/hugepages -v "$PWD/conf":/opt/bess/bessctl/conf \
+        -v /lib/firmware/intel:/lib/firmware/intel \
+        --device=/dev/vfio/vfio --device=/dev/vfio/119 --device=/dev/vfio/120 \
+        upf-epc-bess:"$(<VERSION)" -grpc-url=0.0.0.0:10514
+
+docker exec -it pktgen ./bessctl run pktgen_cndp
+```
+
+We can monitor if pktgen is sending packets using the following command:
+
+```
+docker exec -it pktgen ./bessctl monitor tc
+```
+
+If we need to stop sending packets at some point use the following command:
+
+```
+docker exec -it pktgen ./bessctl daemon reset
+```
+
+### CNDP OMEC-UPF multiple worker threads setup
+
+Modify OMEC UPF and CNDP configuration files to support multiple worker threads. Each thread will run UPF pipeline in a different core.
+
+To test multiple worker thread, we need to use CNDP jsonc file with appropriate configuration. An example CNDP jsonc file for 4 worker threads is in [cndp_upf_4worker.jsonc](../conf/cndp_upf_4worker.jsonc). Follow below steps to configure OMEC-UPF pipeline using 4 worker threads which runs on 5 cores (4 BESS worker threads and 1 main thread).
+
+1) Update [upf.json](../conf/upf.json) to set number if worker threads as 4. `"workers": 4,`
+
+2) Update [upf.json](../conf/upf.json) to use appropriate CNDP jsonc file with required number of lports (netdev/qid pair).
+
+`"cndp_jsonc_file": "conf/cndp_upf_4worker.jsonc"`
+
+3) Modify the script [docker_setup.sh](../scripts/docker_setup.sh) and update the function `move_ifaces()` in condition `if [ "$mode" == 'cndp' ]`.
+
+Update `num_q` value same as number of worker threads
+
+Update `start_q_idx` to choose the start queue index to receive n/w packets.
+
+4) Modify the script  [docker_setup.sh](../scripts/docker_setup.sh) to assign 5 cores (4 worker thread, 1 main thread) to run BESS UPF pipeline. To get better performance (optional step), assign `cpuset-cpus` to cores ids same as queue ids used to receive n/w packets.
+
+```
+diff --git a/scripts/docker_setup.sh b/scripts/docker_setup.sh
+index 9058839..7b4ef76 100755
+--- a/scripts/docker_setup.sh
++++ b/scripts/docker_setup.sh
+@@ -102,9 +102,9 @@ function move_ifaces() {
+                fi
+                if [ "$mode" == 'cndp' ]; then
+                        # num queues
+-                       num_q=1
++                       num_q=4
+                        # start queue index
+-                       start_q_idx=22
++                       start_q_idx=6
+                        # RSS using TC filter
+                        setup_tc "${ifaces[$i]}" $num_q $start_q_idx
+                fi
+@@ -218,7 +218,7 @@ fi
+ # Run bessd
+ docker run --name bess -td --restart unless-stopped \
+-       --cpuset-cpus=12-13 \
++       --cpuset-cpus=6-10 \
+        --ulimit memlock=-1 -v /dev/hugepages:/dev/hugepages \
+        -v "$PWD/conf":/opt/bess/bessctl/conf \
+        --net container:pause \
+```
```

Comment:
I think these two items can be combined because the `diff` in 4) includes also what is described in 3), correct?